cpp/unreal_plugins/UrdfRobot/Source/UrdfRobot/UrdfJointComponent.cpp

```diff
@@ -0,0 +1,487 @@
+    {"Six",        {"add_to_angular_velocity_target",    {   -0.1,  0.0,  0.0}}},
+    {"Seven",      {"add_to_linear_velocity_target",     {    0.1,  0.0,  0.0}}},
+    {"Eight",      {"add_to_linear_velocity_target",     {   -0.1,  0.0,  0.0}}},
+    {"Nine",       {"add_to_angular_orientation_target", {    0.0,  0.0,  2.0}}},
+    {"Zero",       {"add_to_angular_orientation_target", {    0.0,  0.0, -2.0}}},
+    {"Underscore", {"add_to_linear_position_target",     {    2.0,  0.0,  0.0}}},
+    {"Equals",     {"add_to_linear_position_target",     {   -2.0,  0.0,  0.0}}}
+};
+
+UUrdfJointComponent::UUrdfJointComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    // UInputActionComponent
+    input_action_component_ = CreateDefaultSubobject<UInputActionComponent>(Unreal::toFName("input_action_component"));
+    SP_ASSERT(input_action_component_);
+    // Need to explicitly set this up so that the component hierarchy is well-defined.
+    input_action_component_->SetupAttachment(this);
+}
+
+UUrdfJointComponent::~UUrdfJointComponent()
+{
+    SP_LOG_CURRENT_FUNCTION();
+
+    JointType = EJointType::Invalid;
+    JointControlType = EJointControlType::NotActuated;
+    EnableKeyboardControl = false;
+    //LinearTranslationOffset = FVector::ZeroVector; // TODO (MR): support linear translation offsets
+
+    parent_static_mesh_component_ = nullptr;
+    child_static_mesh_component_ = nullptr;
+
+    SP_ASSERT(input_action_component_);
+    input_action_component_ = nullptr;
+}
+
+void UUrdfJointComponent::BeginPlay()
+{
+    UPhysicsConstraintComponent::BeginPlay();
+
+    const std::map<std::string, std::pair<std::string, std::vector<double>>> input_actions = INPUT_ACTIONS;
+    input_action_component_->bindInputActions(input_actions);
+    input_action_component_->apply_input_action_func_ = [this, input_actions](const std::string& key) -> void {
+        if (EnableKeyboardControl) {
+            std::pair<std::string, std::vector<double>> action_component = input_actions.at(key);
+            // only assert if we're not in the editor
+            bool assert_if_action_is_inconsistent_with_joint = !WITH_EDITOR;
+            applyAction(action_component.first, action_component.second, assert_if_action_is_inconsistent_with_joint);
+        }
+    };
+}
+
+std::map<std::string, ArrayDesc> UUrdfJointComponent::getActionSpace() const
+{
+    std::map<std::string, ArrayDesc> action_space;
+
+    if (JointType == EJointType::Invalid || JointType == EJointType::Fixed) {
+        return action_space;
+    }
+
+    ArrayDesc array_desc;
+    array_desc.low_ = std::numeric_limits<double>::lowest();
+    array_desc.high_ = std::numeric_limits<double>::max();
+    array_desc.shape_ = {3};
+    array_desc.datatype_ = DataType::Float64;
+
+    std::string separator = ".";
+    std::string action_name = Unreal::toStdString(GetName()) + separator;
+
+    switch (JointControlType) {
+        case EJointControlType::Position:
```

Comment:
You need to support `set` actions here. And I think `Position`, `PositionAndVelocity`, and `Velocity` all need to be handled in distinct cases because, for example, you're not adding `add_to_angular_velocity_target` even when the control type is `PositionAndVelocity`.

Reply:
Updated the code.